### PR TITLE
RavenDB-8924 add support for regex queries

### PR DESCRIPTION
**Pull request by talweiss1982**
* Had to change index operation to contain a ref to the Index
* Added per index Regex cache

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/ConcurrentLruRegexCache.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+        private static void Updatetimestamp(ConcurrentLruRegexCacheNode result)
+        {
+            var timestamp = Stopwatch.GetTimestamp();
+
+            var resultTimestamp = result.Timestamp;
+            // here we want to avoid updating this using Interlocked on a highly
+            // used query, so we limit it to once per 10,000 ticks, which should
+            // still be good enough if we need to start evicting
+            while (resultTimestamp - timestamp > 10_000)
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/ConcurrentLruRegexCache.cs:48]: Actually need to update the timestamp as well.

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 2]
```diff
@@ -2593,7 +2597,26 @@ public void LowMemoryOver()
         {
         }
 
-        internal SpatialField GetOrAddSpatialField(string name)
+        internal QueryBuilderFactories GetQueryBuilderFactories()
+        {
+            //We don't want to always generate this object but if we generate it mutiple times at startup it isn't really importent.
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs:2602]: Generate it always, it is better to have the creation than the branch

File: src/Raven.Server/Documents/Indexes/ConcurrentLruRegexCache.cs (hunk 1 above)
Comment (talweiss1982) [reply src/Raven.Server/Documents/Indexes/ConcurrentLruRegexCache.cs:48]: i'm not following which time stamp you refer to here,
do you mean to add
timestamp = Stopwatch.GetTimestamp();
to the while?

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/ConcurrentLruRegexCache.cs:48]: yes

**Pull request merged**